### DISCUSSION TITLE
[Tree Shake] Next.js emit unused code into the `main` bundle

### DISCUSSION THREAD

**Discussion by SukkaW** [vercel/next.js#35907] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.4.0: Mon Feb 21 20:34:37 PST 2022; root:xnu-8020.101.4~2/RELEASE_X86_64
    Binaries:
      Node: 16.14.0
      npm: 8.5.1
      Yarn: 1.22.18
      pnpm: 6.32.4
    Relevant packages:
      next: 12.1.5-canary.1
      react: 18.0.0
      react-dom: 18.0.0

### How are you deploying your application? (if relevant)

next build / export

### Describe the Bug

I notice the `main` bundle emitted by `next build` is a bit heavy, so I use `@next/bundle-analyzer` to analyze my bundle:

<img width="1871" alt="image" src="https://user-images.githubusercontent.com/40715044/161738475-77958f1a-aecf-4049-a97d-c733c3e1f334.png">

And there are many features I don't use at all that are included in the bundle:

**next/dist/shared/amp.js**

I don't use the AMP feature at all. So I take a look at the source code. `shared/lib/amp.ts` has 2 named export, `useAmp` and `isInAmpMode`:

https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/shared/lib/amp.ts#L4-L15

`isInAmpMode` is then imported by `shared/lib/head.tsx`:

https://github.com/vercel/next.js/blob/fbfe430cf0d21513eadb7be7e9c05a6ac6d5048a/packages/next/shared/lib/head.tsx#L5

I do use `next/head` component, so it is very reasonable that `isInAmpMode` is included in my bundle. However, `useAmp` is also included in the final bundle:

<img width="1259" alt="image" src="https://user-images.githubusercontent.com/40715044/161739426-43eec14a-c1b9-45b0-85d1-06a600fd1596.png">

**next/dist/client/script.js**

I don't use `next/script` component at all. And the `scriptLoader` property of the `__NEXT_DATA__` is also an empty array:

<img width="815" alt="image" src="https://user-images.githubusercontent.com/40715044/161739866-9e8826fc-3a2e-47df-a254-ac540f374b41.png">

So I take a look at the source code again. `client/script.tsx` has a default export and a named export `initScriptLoader`:

https://github.com/vercel/next.js/blob/fbfe430cf0d21513eadb7be7e9c05a6ac6d5048a/packages/next/client/script.tsx#L132-L134

...which is then imported by `client/index.tsx`:

https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/client/index.tsx#L254-L257

However, the entire `next/dist/client/script.js` is still included in the final `main` bundle:

<img width="1233" alt="image" src="https://user-images.githubusercontent.com/40715044/161740971-dd98de47-98ce-48ad-99eb-7c948c9b3c7e.png">

> As you can see there are keywords like `text/partydown`, `afterInteractive` and `lazyOnLoad`.

**next/dist/client/streaming/vitals.js**

I do use `reportWebVitals` named export on my custom `_app`, but I don't use `unstable_useWebVitalsReport`. So once again, I take a look at the source code of `next.js`.

There are tons of named export at `client/streaming/vitals.ts`:

https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/client/streaming/vitals.ts#L5
https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/client/streaming/vitals.ts#L10-L12
https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/client/streaming/vitals.ts#L19-L22
https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/client/streaming/vitals.ts#L24

`flushBufferedVitalsMetrics` and `trackWebVitalMetric` are imported by `client/index.tsx`:

https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/client/index.tsx#L36-L39

... which is used by `<Root />` component:

https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/client/index.tsx#L1032-L1036

... and `hydrate` function:

https://github.com/vercel/next.js/blob/0eb9f7e76d0d7ec6aa7d85d3eaf57b151f381fb3/packages/next/client/index.tsx#L323

However, `useWebVitalsReport` is also included in my final bundle:

<img width="1244" alt="image" src="https://user-images.githubusercontent.com/40715044/161742736-14a8aa7b-e9cb-4b8f-9fbb-9d7a62b8746b.png">

**next/dist/client/streaming/refresh.js**

I am pretty sure I don't use `useRefreshRoot` at all. And Next.js itself doesn't uses `useRefreshRoot` either. However, it is still included in the final bundle:

<img width="1251" alt="image" src="https://user-images.githubusercontent.com/40715044/161743052-0cc4b934-bf42-4d49-9f9b-fd29da91e0c8.png">

### Expected Behavior

All those unused codes I mentioned above should not be included in the final `main` bundle.

### To Reproduce

I introduce `@next/bundle-analyzer` into another bare bone Next.js reproduction (the one I used in https://github.com/vercel/next.js/issues/35286), and all those unused codes are included in the `main` bundle:

<img width="1874" alt="image" src="https://user-images.githubusercontent.com/40715044/161743937-e54c5e3f-c1c2-48df-94e9-99b30d3e6b6f.png">

**Comment by styfle** [comment]
Thanks for reporting @SukkaW!

The amp bug has been fixed in #37383.

The others are still being investigated. If you have a solution, please submit a PR and we'll take a look, thanks!
